File: src/main/java/chiloven/lukosbot2/commands/IpCommand.java
package chiloven.lukosbot2.commands;

import chiloven.lukosbot2.core.command.CommandSource;
import chiloven.lukosbot2.util.feature.IpService;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import lombok.extern.log4j.Log4j2;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Service;

import java.io.IOException;

import static chiloven.lukosbot2.util.brigadier.builder.LiteralArgumentBuilder.literal;
import static chiloven.lukosbot2.util.brigadier.builder.RequiredArgumentBuilder.argument;

@Service
@ConditionalOnProperty(
        prefix = "lukos.commands.control",
        name = "ip",
        havingValue = "true",
        matchIfMissing = true
)
@Log4j2
public class IpCommand implements BotCommand {
    public static final IpService IS = new IpService();

    @Override
    public String name() {
        return "ip";
    }

    @Override
    public String description() {
        return "查询 IP 信息";
    }

    @Override
    public String usage() {
        return """
                用法：
                /ip <ip_address> # 查询 IP 地址，支持 v4 和 v6
                示例：
                /ip 1.1.1.1
                /ip 2606:4700:4700::1111
                """;
    }

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
                literal(name())
                        .executes(ctx -> {
                            ctx.getSource().reply(usage());
                            return 1;
                        })
                        .then(argument("ip", StringArgumentType.greedyString())
                                .executes(ctx -> {
                                    try {
                                        ctx.getSource().reply(
                                                IS.getIpInfo(StringArgumentType.getString(ctx, "ip")).toString()
                                        );
                                        return 1;
                                    } catch (IOException e) {
                                        log.warn("Failed to fetch ip information.", e);
                                        ctx.getSource().reply(e.getMessage());
                                        return 0;
                                    }
                                })
                        )
        );
    }
}

File: src/main/java/top/chiloven/lukosbot2/commands/impl/ip/IpCommand.kt
package top.chiloven.lukosbot2.commands.impl.ip

import org.apache.logging.log4j.LogManager
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Service
import top.chiloven.lukosbot2.commands.definition.ArgType
import top.chiloven.lukosbot2.commands.definition.CommandInvocation
import top.chiloven.lukosbot2.commands.definition.bridge.SpecBotCommand
import top.chiloven.lukosbot2.commands.definition.dsl.arg
import top.chiloven.lukosbot2.commands.definition.dsl.botCommand
import top.chiloven.lukosbot2.commands.impl.ip.IpQueryResult.IpQueryException

@Service
@ConditionalOnProperty(
    prefix = "lukos.commands.control",
    name = ["ip"],
    havingValue = "true",
    matchIfMissing = true
)
class IpCommand(
    private val ipQueryService: IpQueryService
) : SpecBotCommand() {

    private val log = LogManager.getLogger(IpCommand::class.java)

    override fun spec() = botCommand("ip") {
        description = "查询 IP 信息"

        argv {
            positional("ip", ArgType.StringType) {
                required = true
                description = "IP 地址"
            }

            option("providers") {
                names = listOf("-p", "--provider", "--providers")
                type = ArgType.StringType
                splitBy = ","
                description = "指定数据源；多个用逗号分隔"
            }

            execute { args ->
                runQuery(
                    ip = args.get("ip"),
                    providers = args.getOrNull<List<String>>("providers") ?: emptyList()
                )
            }
        }

        syntax("按默认数据源优先级查询", arg("ip_address"))
        param("ip_address", "IP 地址（IPv4 / IPv6）")
        example(
            "ip 1.1.1.1",
            "ip 2606:4700:4700::1111",
            "ip --provider=ipquery 1.1.1.1",
            "ip --provider=ipquery,ipsb 1.1.1.1"
        )
    }

    private fun CommandInvocation.runQuery(ip: String, providers: List<String>) {
        try {
            val result = ipQueryService.query(
                ip = ip,
                requestedProviders = providers
            )
            reply(result.toDisplayText())
        } catch (e: IllegalArgumentException) {
            reply(e.message ?: "参数错误，请检查 IP 地址或数据源名称")
        } catch (e: IpQueryException) {
            log.warn("IP query failed. ip={}", e.ip, e)
            reply(e.message ?: "IP 查询失败，请稍后重试")
        } catch (e: Exception) {
            log.warn("IP query unexpected failure. ip={}", ip, e)
            reply("IP 查询失败：${e.message ?: "未知错误"}")
        }
    }

}
